Sources/Kafka/KafkaConsumer.swift

```diff
@@ -331,33 +352,112 @@ public final class KafkaConsumer: Sendable, Service {
+                return
+            }
+        }
+    }
+
+    /// Run loop polling Kafka for new consumer messages.
+    private func messageRunLoop() async throws {
+        while !Task.isCancelled {
+            let nextAction = self.stateMachine.withLockedValue { $0.nextConsumerPollLoopAction() }
+            switch nextAction {
+            case .pollForAndYieldMessages(let client, let source):
+                // Poll for new consumer messages.
+                let messageResults = self.batchConsumerPoll(client: client)
+                if messageResults.isEmpty {
+                    self.stateMachine.withLockedValue { $0.waitForNewMessages() }
+                } else {
+                    let yieldResult = source.yield(contentsOf: messageResults)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
+                    case .dropped:
+                        return
                     }
                 }
-                try await Task.sleep(for: self.configuration.pollInterval)
-            case .pollWithoutYield(let client):
-                // Ignore poll result.
-                // We are just polling to serve any remaining events queued inside of `librdkafka`.
-                // All remaining queued consumer messages will get dropped and not be committed (marked as read).
-                _ = client.eventPoll()
+            case .pollForMessagesIfAvailable(let client, let source):
+                let messageResults = self.batchConsumerPoll(client: client)
+                if messageResults.isEmpty {
+                    // Still no new messages, so sleep.
+                    try await Task.sleep(for: self.configuration.pollInterval)
+                } else {
+                    // New messages were produced to the partition that we previously finished reading.
+                    let yieldResult = source.yield(contentsOf: messageResults)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
+                    case .dropped:
+                        return
+                    }
+                }
+            case .suspendPollLoop:
                 try await Task.sleep(for: self.configuration.pollInterval)
             case .terminatePollLoop:
                 return
             }
         }
     }
 
+    /// Read `maxMessages` consumer messages from Kafka.
+    ///
+    /// - Parameters:
+    ///     - client: Client used for handling the connection to the Kafka cluster.
+    ///     - maxMessages: Maximum amount of consumer messages to read in this invocation.
+    private func batchConsumerPoll(
+        client: RDKafkaClient,
+        maxMessages: Int = 100
```

Comment:
We probably should provide this argument from poll based on backpressure. Currently, we have 50 messages high watermark as default value for backpressure, so we may break backpressure as twice if read 100 messages and enqueue them to stream. 
Not sure what is right here but maybe we can call it with one of those high or low watermark limits?

Reply:
Yeah we should probably set this to the difference of `high - low`.